smartsim/_core/mli/infrastructure/storage/backbonefeaturestore.py

```diff
@@ -0,0 +1,399 @@
+        """Convert the event to a string
+
+        :returns: a string representation of this instance"""
+        return f"{str(super())}|{self.descriptor}|{self.key}"
+
+
+class EventPublisher(t.Protocol):
+    """Core API of a class that publishes events"""
+
+    def send(self, event: EventBase) -> int:
+        """The send operation"""
+
+
+class EventBroadcaster:
+    """Performs fan-out publishing of system events"""
+
+    def __init__(
+        self,
+        backbone: BackboneFeatureStore,
+        channel_factory: t.Optional[t.Callable[[str], CommChannelBase]] = None,
+        buffer_size: int = 100,
+    ) -> None:
+        """Initialize the EventPublisher instance
+
+        :param backbone: the MLI backbone feature store
+        :param channel_factory: factory method to construct new channel instances
+        :param buffer_size: Maximum number of events to store before discarding.
+        Discards oldest events first.
+        """
+        self._backbone = backbone
+        """The backbone feature store used to retrieve consumer descriptors"""
+        self._channel_factory = channel_factory
+        """A factory method used to instantiate channels from descriptors"""
+        self._channel_cache: t.Dict[str, t.Optional[CommChannelBase]] = defaultdict(
+            lambda: None
+        )
+        """A mapping of instantiated channels that can be re-used. Automatically 
+        calls the channel factory if a descriptor is not already in the collection"""
+        self._event_buffer: t.Deque[bytes] = deque(maxlen=buffer_size)
+        """A buffer for storing events when a consumer list is not found. Buffer size
+        can be fixed by passing the `buffer_size` parameter."""
+        self._num_discards: int = 0
+        """Number of messages discarded from the buffer since the last successful
+        broadcast"""
+        self._buffer_size: int = buffer_size
+        """Maximum size of the event buffer"""
+        self._descriptors: t.Set[str]
+        """Stores the most recent list of broadcast consumers. Updated automatically
+        on each broadcast"""
+
+    @property
+    def num_buffered(self) -> int:
+        """Return the number of events currently buffered to send"""
+        return len(self._event_buffer)
+
+    @property
+    def num_discarded(self) -> int:
+        """Return the number of events discarded from the buffer since the last send"""
+        return self._num_discards
+
+    def _save_to_buffer(self, event: EventBase) -> None:
+        """Places a serialized event in the buffer until full. Continues adding events
+        to the buffer using a first-in, first-discarded strategy.
+
+        :param event: The event to serialize and buffer"""
+        if len(self._event_buffer) == self._buffer_size:
+            self._event_buffer.popleft()
+            self._num_discards += 1
+
+        try:
+            event_bytes = bytes(event)
+            self._event_buffer.append(event_bytes)
```

Comment:
I don't think this is preventable, just want to draw some attention that there will be copies here and I should performance check it in the future!

Reply:
I profiled up to a buffer size of 1 million events. That would far exceed my buffering expectations for a single node (by 2 orders of magnitude), but still only results in ~200MB of allocations. 

I think this should alleviate your concern, here?

```py
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
   814     99.6 MiB     99.6 MiB           1   @pytest.mark.skipif(DISABLE_PROFILING != 0, reason=profile_reason)
   815                                         @memory_profiler.profile
   816                                         def test_profile_event_buffer_growth(
   817                                             test_dir: str, monkeypatch: pytest.MonkeyPatch
   818                                         ) -> None:
   819                                             """
   820                                             Execute a test to memory profile when a massive event buffer is used.
   821                                         
   822                                             :param test_dir: pytest fixture automatically generating unique working
   823                                             directories for individual test outputs"""
   824                                             # pytest.mark.skipif(DISABLE_PROFILING, "do not profile during test runs")
   825                                         
   826     99.6 MiB      0.0 MiB           1       storage_path = pathlib.Path(test_dir) / "features"
   827     99.6 MiB      0.0 MiB           1       storage_path.mkdir(parents=True, exist_ok=True)
   828                                         
   829     99.6 MiB      0.0 MiB           1       mock_storage = {}
   830                                         
   831                                             # note: file-system descriptors are just paths
   832     99.6 MiB      0.0 MiB           1       target_descriptor = str(storage_path / "test-consumer")
   833     99.6 MiB      0.0 MiB           1       buffer_size = 1000000
   834                                         
   835     99.6 MiB      0.0 MiB           1       backbone = BackboneFeatureStore(mock_storage)
   836     99.6 MiB      0.0 MiB           2       publisher = EventBroadcaster(
   837     99.6 MiB      0.0 MiB           1           backbone,
   838     99.6 MiB      0.0 MiB           1           channel_factory=FileSystemCommChannel.from_descriptor,
   839     99.6 MiB      0.0 MiB           1           buffer_size=buffer_size,
   840                                             )
   841                                         
   842    309.1 MiB      0.0 MiB           2       with monkeypatch.context() as patch:
   843                                         
   844                                                 # cause buffer to grow to massive size by blocking delivery
   845                                                 # patch.setattr(publisher, "_broadcast", lambda: 0)
   846                                         
   847                                                 # backbone.notification_channels = (target_descriptor,)
   848                                         
   849     99.6 MiB      0.0 MiB           1           events = []
   850                                                 # hit max buffer size
   851    309.1 MiB      0.0 MiB     1000001           for i in range(buffer_size):
   852    309.1 MiB    199.0 MiB     1000000               event = OnCreateConsumer(target_descriptor)
   853    309.1 MiB     10.5 MiB     1000000               events.append(event)
   854                                                     # num_sent = publisher.send(event)
   855                                         
   856                                                     # assert num_sent == 0
   857                                                     # assert len(publisher._event_buffer) == i + 1
```

Reply:
This was more of a comment on the bytes conversion in `_save_to_buffer`. It stuck out to me because of my `tobytes()` trauma lol